cmd/juju/interact/pollster.go

```diff
@@ -0,0 +1,515 @@
+}
+
+// SelectVerify queries the user to select from the give list of options, using
+// the custom verifier listed.
+func (p *Pollster) SelectVerify(l List, verify VerifyFunc) (string, error) {
+	if err := listTmpl.Execute(p.out, l); err != nil {
+		return "", err
+	}
+
+	question, err := sprint(selectTmpl, l)
+	if err != nil {
+		return "", errors.Trace(err)
+	}
+	val, err := QueryVerify(question, p.scanner, p.out, p.errOut, verify)
+	if err != nil {
+		return "", errors.Trace(err)
+	}
+	if val == "" {
+		return l.Default, nil
+	}
+	return val, nil
+}
+
+var multiSelectTmpl = template.Must(template.New("").Funcs(
+	map[string]interface{}{"join": strings.Join}).Parse(`
+Select one or more {{.Plural}} separated by commas{{if .Default}} [{{join .Default ", "}}]{{end}}: `[1:]))
+
+// MultiSelect queries the user to select one more answers from the give list of
+// options by entering values delimited by commas (and thus options must not
+// contain commas).
+func (p *Pollster) MultiSelect(l MultiList) ([]string, error) {
+	if err := listTmpl.Execute(p.out, l); err != nil {
+		return nil, err
+	}
+
+	question, err := sprint(multiSelectTmpl, l)
+	if err != nil {
+		return nil, errors.Trace(err)
+	}
+	a, err := QueryVerify(question, p.scanner, p.out, p.errOut, multiVerify(l.Singular, l.Plural, l.Options, l.Default != nil))
+	if err != nil {
+		return nil, errors.Trace(err)
+	}
+	if a == "" {
+		return l.Default, nil
+	}
+	return multiSplit(a), nil
+}
+
+// Enter requests that the user enter a value.  Any value except an empty string
+// is accepted.
+func (p *Pollster) Enter(valueName string) (string, error) {
+	return p.EnterVerify(valueName, nil)
+}
+
+// VerifyFunc is a type that determines whether a value entered by the user is
+// acceptable or not.  If it returns an error, the calling func will return an
+// error, and the other return values are ignored.  If ok is true, the value is
+// acceptable, and that value will be returned by the calling function.  If ok
+// is false, the user will be asked to enter a new value for query.  If ok is
+// false. if errmsg is not empty, it will be printed out as an error to te the
+// user.
+type VerifyFunc func(s string) (ok bool, errmsg string, err error)
+
+// EnterVerify requests that the user enter a value.  Values failing to verify will be
+// rejected with the error message returned by verify.
+func (p *Pollster) EnterVerify(valueName string, verify VerifyFunc) (string, error) {
+	for {
+		s, err := QueryVerify("Enter "+valueName+": ", p.scanner, p.out, p.errOut, verify)
+		if err != nil {
+			return "", errors.Trace(err)
+		}
+		if s != "" {
+			return s, nil
+		}
+		// retry if we get an empty value, no need to print out an error.
+	}
+}
+
+// YN queries the user with a yes no question q (which should not include a
+// question mark at the end).  It uses def as the default answer.
+func (p *Pollster) YN(q string, def bool) (bool, error) {
+	defstring := "(y/N)"
```

Comment:
yeah, that's better. There's times when I really wish default was not a keyword :)